File: problems/8.stack-qeueue/golang/queue.go
package main

type Queue struct {
	in  []int
	out []int
}

func (q Queue) dequeue() int {
	if len(q.out) == 0 {
		q.fillStackOut()
	}

	last := q.out[len(q.in)-1]
	q.out = q.out[:len(q.out)-1]
	return last
}

func (q Queue) peek() int {
	if len(q.out) == 0 {
		q.fillStackOut()
	}

	return q.out[len(q.in)-1]
}

func (q *Queue) enqueue(value int) {
	q.in = append(q.in, value)
}

func (q Queue) fillStackOut() {
	for len(q.in) > 0 {
		last := q.in[len(q.in)-1]
		q.in = q.in[:len(q.in)-1]
		q.out = append(q.out, last)
	}
}

func (q Queue) empty() bool {
	return len(q.in) == 0 && len(q.out) == 0
}

File: ts/8.stack-qeueue/queue.ts
interface QueueWithStacks<T> {
    enqueue: (value: T) => void;
    dequeue: () => T | undefined;
    peek: () => T;
    empty: () => boolean;
}

export class QueueWithStacksImpl<T> implements QueueWithStacks<T> {
    in: T[];
    out: T[];
    constructor() {
        this.out = [];
        this.in = [];
    }

    public enqueue(value: T) {
        this.in.push(value);
    }

    public dequeue() {
        this.fillRemoveStack();

        return this.out.pop();
    }

    public peek() {
        this.fillRemoveStack();

        return this.out[this.out.length - 1];
    }

    public empty() {
        return this.out.length === 0 && this.in.length === 0;
    }

    private fillRemoveStack() {
        if (this.out.length === 0) {
            while (this.in.length > 0) {
                this.out.push(this.in.pop() as T);
            }
        }
    }
}
